### PR TITLE
[project] Migrate into new API

### PR DESCRIPTION
**Pull request by gan-odoo**
Task : https://www.odoo.com/web#id=20786&view_type=form&model=project.task&action=327
Pad : https://pad.odoo.com/p/r.7TGralhxM5oEeddi

### CODE REVIEW

File: addons/project/models/project.py [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+                              ('pending','Pending'),
+                              ('close','Closed')],
+                              'Status', required=True, copy=False, default='open')
+    doc_count = fields.Integer(compute='_compute_get_attached_docs', string="Number of documents attached")
+    date_start = fields.Date('Start Date')
+    date = fields.Date('Expiration Date', select=True, track_visibility='onchange')
+
+    _defaults = {
+        'use_tasks': True
+    }
```
Comment (WHPODOO) [inline addons/project/models/project.py]: _defaults ? :laughing: :laughing: :laughing: :laughing: :laughing: :laughing:

File: addons/project/models/project.py [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+        fold = {}
+        for stage in stages:
+            fold[stage.id] = stage.fold
+        return result, fold
+
+    _group_by_full = {
+        'stage_id': _read_group_stage_ids,
+    }
+
+    active = fields.Boolean('Active', default=True)
```
Comment (WHPODOO) [inline addons/project/models/project.py]: No need to give string

File: addons/project/models/project.py [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+    name = fields.Char('Task Title', track_visibility='onchange', required=True, select=True)
+    description = fields.Html('Description')
```
Comment (WHPODOO) [inline addons/project/models/project.py]: No need to give string

File: addons/project/models/project.py [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+    priority = fields.Selection([('0','Normal'), ('1','High')], 'Priority', select=True, default=0)
```
Comment (WHPODOO) [inline addons/project/models/project.py]: No need to give string

File: addons/project/models/project.py [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+    sequence = fields.Integer('Sequence', select=True, default=10, help="Gives the sequence order when displaying a list of tasks.")
```
Comment (WHPODOO) [inline addons/project/models/project.py]: No need to give string

File: addons/project/models/account_analytic.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = 'account.analytic.account'
+    _description = 'Analytic Account'
+
+    use_tasks = fields.Boolean('Tasks', help="Check this box to manage internal activities through this project")
+    company_uom_id = fields.Many2one(related='company_id.project_time_mode_id', string="Company UOM")
+
+
+    @api.model
```
Comment (atp-odoo) [inline addons/project/models/account_analytic.py]: useless decorator. decorator is not required...

File: addons/project/models/account_analytic.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    def _trigger_project_creation(self, vals):
+        '''
+        This function is used to decide if a project needs to be automatically created or not when an analytic account is created. It returns True if it needs to be so, False otherwise.
+        '''
+        return vals.get('use_tasks') and not 'project_creation_in_progress' in self.env.context
+
+    def project_create(self, vals):
```
Comment (atp-odoo) [inline addons/project/models/account_analytic.py]: are you sure decorator not required ?

File: addons/project/models/account_analytic.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        Project = self.env['project.project']
+        projects = Project.search([('analytic_account_id','=', self.id)])
+        if not projects and self._trigger_project_creation(vals):
+            project_values = {
+                'name': vals.get('name'),
+                'analytic_account_id': self.id,
+                'use_tasks': True,
+            }
+            return Project.create(project_values)
+        return False
```
Comment (atp-odoo) [inline addons/project/models/account_analytic.py]: no need to return False.

File: addons/project/models/account_analytic.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    def write(self, vals):
+        vals_for_project = vals.copy()
+        if not vals.get('name'):
+            vals_for_project['name'] = self.name
+        self.project_create(vals_for_project)
+        return super(AccountAnalyticAccount, self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        count_tasks = self.env['project.task'].search([('project_id.analytic_account_id', 'in', self.ids)], count=True)
```
Comment (atp-odoo) [inline addons/project/models/account_analytic.py]: use search_count method

File: addons/project/models/partner.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class Partner(models.Model):
+
+    @api.multi
+    def _compute_task_count(self):
```
Comment (atp-odoo) [inline addons/project/models/partner.py]: define this after field declaration ....

File: addons/project/models/project.py [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+    _order = 'sequence'
+
+    def _get_default_project_ids(self):
+        default_project_id = self.env.context.get('default_project_id')
+        return [default_project_id] if default_project_id else None
+
+    name = fields.Char('Stage Name', required=True, translate=True)
+    description = fields.Text('Description', translate=True)
+    sequence = fields.Integer('Sequence', default=1)
+    project_ids = fields.Many2many('project.project', 'project_task_type_rel', 'type_id', 'project_id', 'Projects', default=_get_default_project_ids)
```
Comment (atp-odoo) [inline addons/project/models/project.py]: cannot we use lambda instead of call function ?

File: addons/project/models/project.py [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+                "   is activated, portal users see the followed tasks or issues.")
+    state = fields.Selection([('draft','New'),
+                              ('open','In Progress'),
+                              ('cancelled', 'Cancelled'),
+                              ('pending','Pending'),
+                              ('close','Closed')],
+                              'Status', required=True, copy=False, default='open')
+    doc_count = fields.Integer(compute='_compute_get_attached_docs', string="Number of documents attached")
+    date_start = fields.Date('Start Date')
+    date = fields.Date('Expiration Date', select=True, track_visibility='onchange')
```
Comment (atp-odoo) [inline addons/project/models/project.py]: change field name with keep oldname and make separate commit...

File: addons/project/models/account_analytic.py (hunk 7 above)
Comment (gan-odoo) [inline addons/project/models/account_analytic.py]: @atp-odoo Yes, decorator not required because this method is call from internal method.

File: addons/project/models/project.py [hunk 13]
```diff
@@ -0,0 +1,633 @@
...
+    type_ids = fields.Many2many('project.task.type', 'project_task_type_rel', 'project_id', 'type_id', 'Tasks Stages', states={'close':[('readonly',True)], 'cancelled':[('readonly',True)]})
+    task_count = fields.Integer(compute='_compute_task_count', string="Tasks")
+    task_ids = fields.One2many('project.task', 'project_id',
+                                domain=[('stage_id.fold', '=', False)])
+    color = fields.Integer('Color Index')
+    user_id = fields.Many2one('res.users', 'Project Manager', default=lambda self: self.env.user)
+    alias_id = fields.Many2one('mail.alias', 'Alias', ondelete="restrict", required=True,
+                                help="Internal email associated with this project. Incoming emails are automatically synchronized "
+                                     "with Tasks (or optionally Issues if the Issue Tracker module is installed).")
+    alias_model = fields.Selection(_alias_models, "Alias Model", select=True, required=True, default='project.task',
```
Comment (atp-odoo) [inline addons/project/models/project.py]: select=True deprecated.. use index=True

File: addons/project/models/project.py (hunk 12 above)
Comment (atp-odoo) [inline addons/project/models/project.py]: Use index instead of select

File: addons/project/models/project.py (hunk 1 above)
Comment (atp-odoo) [inline addons/project/models/project.py]: _defaults  ?????????????????

File: addons/project/models/project.py [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+    def _compute_task_count(self):
+        task_data = self.env['project.task'].read_group([('project_id', 'in', self.ids),'|', ('stage_id.fold', '=', False), ('stage_id', '=', None)], ['project_id'], ['project_id'])
+        mapped_data = {task['project_id'][0]: task['project_id_count'] for task in task_data}
+        for project in self:
+            project.task_count = mapped_data.get(project.id)
+
+    @api.multi
+    def _compute_get_attached_docs(self):
+        Attachment = self.env['ir.attachment']
+        attachment_data = Attachment.read_group([('res_model', '=', 'project.project'), ('res_id', 'in', self.ids)], ['res_id'], ['res_id'])
```
Comment (atp-odoo) [inline addons/project/models/project.py]: why not search here instead of read_group ?

File: addons/project/models/project.py [hunk 15]
```diff
@@ -0,0 +1,633 @@
...
+        project_attachments = {attachment['res_id']: attachment['res_id_count'] for attachment in attachment_data}
+        Task = self.env['project.task']
+        for project in self:
+            tasks = Task.search([('project_id', '=', project.id)])
+            task_attachments = Attachment.search([('res_model', '=', 'project.task'), ('res_id', 'in', tasks.ids)], count=True)
+            project.doc_count = project_attachments.get(project.id, 0) + (task_attachments and task_attachments.id or 0)
+
+    @api.constrains('date_start', 'date')
+    def _check_dates(self):
+        if self.date_start and self.date and self.date_start > self.date:
```
Comment (atp-odoo) [inline addons/project/models/project.py]: is it work self.date_start when self contain multiple recordset ?  @api.constraints without decorator take default @api.multi . so iteration required.

File: addons/project/models/account_analytic.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+            vals['child_ids'] = []
+        analytic_account = super(AccountAnalyticAccount, self).create(vals)
+        analytic_account.project_create(vals)
+        return analytic_account
+
+    @api.multi
+    def write(self, vals):
+        vals_for_project = vals.copy()
+        if not vals.get('name'):
+            vals_for_project['name'] = self.name
```
Comment (mdi-odoo) [inline addons/project/models/account_analytic.py]: Trying to read or write a field on multiple records will raise an error.

File: addons/project/models/partner.py [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+class Partner(models.Model):
+
+    """ Inherits partner and adds Tasks information in the partner form """
+    _inherit = 'res.partner'
+
+    task_ids = fields.One2many('project.task', 'partner_id', string='Tasks')
+    task_count = fields.Integer(compute='_compute_task_count', string='# Tasks')
+
+    @api.multi
+    def _compute_task_count(self):
```
Comment (mdi-odoo) [inline addons/project/models/partner.py]: No need of decorator .

File: addons/project/models/project.py [hunk 18]
```diff
@@ -0,0 +1,653 @@
...
+                              ('close','Closed')],
+                              'Status', required=True, copy=False, default='open')
+    doc_count = fields.Integer(compute='_compute_get_attached_docs', string="Number of documents attached")
+    date_start = fields.Date('Start Date')
+    date_end = fields.Date('Expiration Date', index=True, oldname='date', track_visibility='onchange')
+    use_tasks = fields.Boolean(related='analytic_account_id.use_tasks', default=True)
+
+
+    @api.multi
+    def _compute_task_count(self):
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: No need of decorator .

File: addons/project/models/project.py [hunk 19]
```diff
@@ -0,0 +1,653 @@
...
+        task_data = self.env['project.task'].read_group([('project_id', 'in', self.ids),'|', ('stage_id.fold', '=', False), ('stage_id', '=', None)], ['project_id'], ['project_id'])
+        mapped_data = {task['project_id'][0]: task['project_id_count'] for task in task_data}
+        for project in self:
+            project.task_count = mapped_data.get(project.id)
+
+    @api.multi
+    def _compute_get_attached_docs(self):
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: Same here .

File: addons/project/models/project.py [hunk 20]
```diff
@@ -0,0 +1,653 @@
...
+    @api.constrains('date_start', 'date_end')
+    def _check_dates(self):
+        for project in self:
+            if project.date_start and project.date_end and project.date_start > project.date_end:
+                raise UserError(_('Error! project start-date must be lower than project end-date.'))
+
+    def map_tasks(self, project):
+        """ copy and map tasks from old to new project """
+        map_task_id = {}
+        for task in self.tasks:
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: Trying to read or write a field on multiple records will raise an error.

File: addons/project/models/project.py [hunk 21]
```diff
@@ -0,0 +1,653 @@
...
+        for project in self:
+            if project.tasks:
+                raise UserError(_('You cannot delete a project containing tasks. You can either delete all the project\'s tasks and then delete the project or simply deactivate the project.'))
+            elif project.alias_id:
+                alias_ids.append(project.alias_id.id)
+            if not project.analytic_account_id.line_ids:
+                analytic_account_id.add(project.analytic_account_id.id)
+        res = super(Project, self).unlink()
+        self.env['mail.alias'].search([('id', 'in', alias_ids)]).unlink()
+        self.env['account.analytic.account'].search([('id', 'in', list(analytic_account_id))]).unlink()
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: Why you need search() method here ? Use set operation on recordset(s) and use that recordset(s) to unlink the record(s).

File: addons/project/models/project.py [hunk 22]
```diff
@@ -0,0 +1,653 @@
...
+        self.remaining_hours = self.planned_hours
+
+    @api.onchange('project_id')
+    def onchange_project_id_wrapper(self):
+        result = self.onchange_project(self.project_id.id)['value']
+        for fname, value in result.iteritems():
+            setattr(self, fname, value)
+
+    @api.multi
+    def onchange_project(self, project_id):
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: Why @api.multi on 'onchange' method ? Make pure @api.onchange method .

Comment (gan-odoo) [inline addons/project/models/project.py]: @mdi-odoo I make this wrapper method because onchange_project() is override from 'project_timesheet' module.

File: addons/project/models/project.py [hunk 23]
```diff
@@ -0,0 +1,653 @@
...
+        values = {'alias_parent_thread_id': project.id, 'alias_defaults': {'project_id': project.id}}
+        project.alias_id.write(values)
+        return project
+
+    @api.multi
+    def write(self, vals):
+        # if alias_model has been changed, update alias_model_id accordingly
+        if vals.get('alias_model'):
+            models = self.env['ir.model'].search([('model', '=', vals.get('alias_model', 'project.task'))])
+            vals.update(alias_model_id=models.id)
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: Use limit = 1 .

File: addons/project/models/project.py [hunk 24]
```diff
@@ -0,0 +1,647 @@
...
+            section_ids.append(section_id)
+        search_domain = []
+        if section_ids:
+            search_domain = [('|')] * (len(section_ids) - 1)
+            for section_id in section_ids:
+                search_domain.append(('project_ids', '=', section_id))
+        search_domain += list(domain)
+        # perform search, return the first found
+        stages = self.env['project.task.type'].search(search_domain, order=order)
+        return stages and stages[0] or False
```
Comment (mdi-odoo) [inline addons/project/models/project.py]: Use limit = 1 .

**Comment by hmo-odoo** [comment]
continue work on PR: https://github.com/odoo/odoo/pull/11041
